src/emulator/taskQueue.ts

```diff
@@ -0,0 +1,428 @@
+  //  - There are tasks within the queue
+  //  - There is space in the dispatch
+  //  - There are tokens available (used for rate limiting)
+
+  dispatchTasks(): void {
+    while (!this.queue.isEmpty() && this.openDispatches.length > 0 && this.tokens >= 1) {
+      const dispatchLocation = this.openDispatches.pop();
+      if (dispatchLocation !== undefined) {
+        const dispatch = this.queue.dequeue();
+
+        dispatch.metadata.lastRunTime = null;
+        dispatch.metadata.currentAttempt = 1;
+        dispatch.metadata.currentBackoff = this.config.retryConfig.minBackoffSeconds;
+        dispatch.metadata.status = TaskStatus.NOT_STARTED;
+        dispatch.metadata.startTime = Date.now();
+
+        this.dispatches[dispatchLocation] = dispatch;
+        this.tokens--;
+      }
+    }
+  }
+
+  // Used for testing
+  setDispatch(dispatches: (EmulatedTask | null)[]): void {
+    this.dispatches = dispatches;
+    const open = [];
+    for (let i = 0; i < this.dispatches.length; i++) {
+      if (dispatches[i] === null) {
+        open.push(i);
+      }
+    }
+    this.openDispatches = open;
+  }
+
+  getDispatch(): (EmulatedTask | null)[] {
+    return this.dispatches;
+  }
+
+  // Updates the status of all tasks that are currently in the task dispatch
+  processDispatch(): void {
+    for (let i = 0; i < this.dispatches.length; i++) {
+      if (this.dispatches[i] !== null) {
+        switch (this.dispatches[i]?.metadata.status) {
+          case TaskStatus.FAILED:
+            this.dispatches[i] = null;
+            this.openDispatches.push(i);
+            this.completedTimes.push(Date.now());
+            this.failedTimes.push(Date.now());
+            break;
+          case TaskStatus.NOT_STARTED:
+            void this.runTask(i);
+            break;
+          case TaskStatus.RETRY:
+            this.handleRetry(i);
+            break;
+          case TaskStatus.FINISHED:
+            this.dispatches[i] = null;
+            this.openDispatches.push(i);
+            this.completedTimes.push(Date.now());
+            break;
+        }
+      }
+    }
+  }
+
+  async runTask(dispatchIndex: number): Promise<void> {
+    if (this.dispatches[dispatchIndex] === null) {
+      throw new Error("Trying to dispatch a nonexistent task");
+    }
+
+    const emulatedTask = this.dispatches[dispatchIndex] as EmulatedTask;
+
+    if (
+      emulatedTask.metadata.lastRunTime !== null &&
+      Date.now() - emulatedTask.metadata.lastRunTime < emulatedTask.metadata.currentBackoff * 1000
+    ) {
+      // Task is not yet ready to run
+      return;
+    }
+
+    emulatedTask.metadata.status = TaskStatus.RUNNING;
+    try {
+      const response = await fetch(emulatedTask.task.httpRequest.url, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          ...emulatedTask.task.httpRequest.headers,
+        },
+        body: JSON.stringify(emulatedTask.task.httpRequest.body),
+      });
```

Comment:
hmm what happens when this response takes extremely long time to respond?

Reply:
There is the `dispatchDeadlineSeconds` parameter on tasks, which can specify how long to wait for a task to respond. Right now the only task parameters that are respected by the emulator are URI and ID. I don't think adding in support for `dispatchDeadlineSeconds` would be too difficult though so I can look into that.